R/github.R

```diff
@@ -147,14 +148,29 @@ use_github <- function(organisation = NULL,
   if (protocol == "ssh") {
     ## [1] push via ssh required for success setting remote tracking branch
     ## [2] to get passphrase from ssh-agent, you must use NULL credentials
-    git2r::push(r, "origin", "refs/heads/master", credentials = credentials)
+    pushed <- tryCatch({
+      git2r::push(r, "origin", "refs/heads/master", credentials = credentials)
+      TRUE
+    }, error = function(e) FALSE)
   } else { ## protocol == "https"
     ## in https case, when GITHUB_PAT is passed as password,
     ## the username is immaterial, but git2r doesn't know that.
     cred <- git2r::cred_user_pass("EMAIL", auth_token)
-    git2r::push(r, "origin", "refs/heads/master", credentials = cred)
+    pushed <- tryCatch({
+      git2r::push(r, "origin", "refs/heads/master", credentials = cred)
+      TRUE
+    }, error = function(e) FALSE)
+  }
+  if (pushed) {
+    git2r::branch_set_upstream(git2r::repository_head(r), "origin/master")
+  } else {
+    ui_todo(c(
+      "Failed to push and set tracking branch.",
+      "This often indicates a problem with git2r and the credentials.",
+      "Try this in the shell, to complete the set up:",
+      "{ui_code('git push --set-upstream origin master')}"
+    ))
```

Comment:
Here's where we are most likely to fail, esp. on Windows, due to git2r credential-finding and using problems. I think it makes sense to catch that and provide the git command that will probably finish the job.

Reply:
This a version of @maelle's suggestion in the issue thread https://github.com/r-lib/usethis/issues/221#issuecomment-432581263

Reply:
Shouldn't we encourage to re-run `use_github()`? That will cause more pain for us in the short-term, but will force us to resolve the problems in the loing-term.

Reply:
Well, no, I think.

I played with various options to dealing with partial setup, but it's hard to account for the various weird scenarios people could show up with. If we fail here, the first push is the only thing that hasn't happened. The GitHub repo exists and 'origin' is configured with that URL. I think it's ok to leave people here. Literally the only thing they need to to is `git push --set-upstream origin master`.

Reply:
If they fail due to ... not a git repo, not on master, working tree dirty, GitHub repo pre-exists, no APT ... we can (and do) tell them what to fix. They fix and call `use_github()` again.

If they fail due to because `gitr2::push()` fails, I don't even know how to help them. Do they need to build git2r from source to link against the correct system libraries? Do they need to setup an ssh agent? Do they need to allow the agent to store their passphrase? Who knows?!?